src/transform.py

```diff
@@ -93,28 +96,30 @@ class PretalxSubmission(BaseModel):
     speakers: list[str]  # We only want the code, not the full info
     submission_type: str
     slug: str
-    track: str | None
+    track: str | None = None
     state: str
     abstract: str
     answers: list[PretalxAnswer] = Field(..., exclude=True)
     tweet: str = ""
     duration: str
 
     level: str = ""
-    delivery: str | None = ""
+    delivery: str = ""
 
     # This is embedding a slot inside a submission for easier lookup later
     room: str | None = None
     start: datetime | None = None
     end: datetime | None = None
 
     # TODO: once we have schedule data then we can prefill those in the code here
+    # These are added after the model is created
     talks_in_parallel: list[str] | None = None
     talks_after: list[str] | None = None
-    next_talk_code: str | None = None
-    prev_talk_code: str | None = None
+    talks_before: list[str] | None = None
+    next_talk: str | None = None
+    prev_talk: str | None = None
 
-    website_url: str | None = None
+    website_url: str
 
     @model_validator(mode="before")
     @classmethod
```

Comment:
Instead of doing a custom extract, can we delegate this to Pydantic?
This theoretically should remove all `if` `else` branching below and will make parsing easier to understand. What do you think?

Reply:
What I mean is something like this. Since we know the data model from Pretix and the data model we want to serve after we could simplify code a lot by utilizing `.parse_raw` method.

Here is the simple example on what I mean:

```Python
from pydantic import BaseModel
from typing import List
import json

class Item(BaseModel):
    name: str
    price: float
    quantity: int

class Order(BaseModel):
    order_id: int
    items: List[Item]


json_data = '''
{
    "order_id": 123,
    "items": [
        {"name": "apple", "price": 1.2, "quantity": 10},
        {"name": "banana", "price": 0.8, "quantity": 5}
    ]
}
'''

order = Order.parse_raw(json_data)

print(order.order_id)
for item in order.items:
    print(f"Item: {item.name}, Price: {item.price}, Quantity: {item.quantity}")
```

Reply:
I'm new to Pydantic, and these tips are really helpful for me. Thanks ❤️

Reply:
I agree. Pydantic can support us in two major ways:
1. Represent and work with data in a structured way (as BaseModels) instead of hacking around with nested dicts and lists.
2. Convert data between JSON strings and Python objects.

If we only want the JSON de-/serialization, having huge `model_validator`s a valid the solution: We can do all the hacky `dict` magic we want, and in the end we get a `BaseModel` instance.

Pydantic also offers some tools to perform transformations in a more structured way:
- [computed_field](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.computed_field) to create new fields based on parsed data
- [exclude](https://docs.pydantic.dev/latest/concepts/fields/#exclude) to exclude fields during serialization
- [aliases](https://docs.pydantic.dev/latest/concepts/fields/#field-aliases) to automatically rename fields
- [Enum parsing](https://docs.pydantic.dev/2.0/usage/types/enums/) to avoid "magic strings"

<details>
  <summary>[Click me] Example from the Pretix API</summary>

  ```python
from enum import Enum

from pydantic import BaseModel, Field, computed_field


class PretixOrderStatus(Enum):
    PAID = "p"
    PENDING = "n"
    EXPIRED = "e"
    CANCELED = "c"


class PretixOrderPosition(BaseModel):
    order_id: str = Field(alias="order")
    item_id: int = Field(alias="item")
    attendee_name: str | None


class PretixOrder(BaseModel):
    id: str = Field(alias="code")
    status: PretixOrderStatus = Field(exclude=True)
    positions: list[PretixOrderPosition]

    @computed_field
    def is_paid(self) -> bool:
        return self.status is PretixOrderStatus.PAID


order_json = """
{
  "code": "ABC01",
  "status": "p",
  "positions": [
    {
      "order": "ABC01",
      "item": 123,
      "attendee_name": "Jane Doe"
    },
    {
      "order": "ABC01",
      "item": 234,
      "attendee_name": "John Doe"
    }
  ]
}
"""

order = PretixOrder.model_validate_json(order_json)
assert order.model_dump_json(indent=2) == """\
{
  "id": "ABC01",
  "positions": [
    {
      "order_id": "ABC01",
      "item_id": 123,
      "attendee_name": "Jane Doe"
    },
    {
      "order_id": "ABC01",
      "item_id": 234,
      "attendee_name": "John Doe"
    }
  ],
  "is_paid": true
}"""
  ```
</details>

I know this is a matter of opinion, feel free to keep the code as it is.

PS: `BaseModel.parse_raw` is [deprecated in Pydantic v2](https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel), please use `BaseModel.model_validate_json` instead 🙂